### ISSUE TITLE
Failed to deploy Azure ML Model 

### ISSUE THREAD

**Issue by santhosh340** [microsoft/MLOpsPython#171]
I am using Azure DevOps pipeline to deploy azure ml model, I am not able to deploy ml model to AKS. FYI error is below.

Response Code: 404
Headers: {'Date': 'Mon, 03 Feb 2020 20:23:37 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Vary': 'Accept-Encoding', 'Request-Context': 'appId=cid-v1:2d2e8e63-272e-4b3c-8598-4ee570a0e70d', 'api-supported-versions': '1.0, 2018-03-01-preview, 2018-11-19', 'x-ms-client-request-id': '32956e8e1d724ae68d37353bcb6f7b3c', 'x-ms-client-session-id': '', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains; preload', 'Content-Encoding': 'gzip'}
Content: b'{"code":"NotFound","statusCode":404,"message":"The specified resource was not found","details":[{"code":"NotFoundComputeResource","message":"Compute Resource $(AKS_COMPUTE_NAME) doesn\'t exist."}]}'
	InnerException None
	ErrorResponse 
{
    "error": {
        "message": "Received bad response from Model Management Service:\nResponse Code: 404\nHeaders: {'Date': 'Mon, 03 Feb 2020 20:23:37 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Vary': 'Accept-Encoding', 'Request-Context': 'appId=cid-v1:2d2e8e63-272e-4b3c-8598-4ee570a0e70d', 'api-supported-versions': '1.0, 2018-03-01-preview, 2018-11-19', 'x-ms-client-request-id': '32956e8e1d724ae68d37353bcb6f7b3c', 'x-ms-client-session-id': '', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains; preload', 'Content-Encoding': 'gzip'}\nContent: b'{\"code\":\"NotFound\",\"statusCode\":404,\"message\":\"The specified resource was not found\",\"details\":[{\"code\":\"NotFoundComputeResource\",\"message\":\"Compute Resource $(AKS_COMPUTE_NAME) doesn\\'t exist.\"}]}'"
    }
}}
Output:  
##[error]Error Code: [1]
##[error]Error: Failed to deploy model(s)
{ code: 1,
  error: undefined,
  stdout: '',
  stderr: 'ERROR: {\'Azure-cli-ml Version\': \'1.0.85\', \'Error\': WebserviceException:\n\tMessage: Received bad response from Model Management Service:\nResponse Code: 404\nHeaders: {\'Date\': \'Mon, 03 Feb 2020 20:23:37 GMT\', \'Content-Type\': \'application/json\', \'Transfer-Encoding\': \'chunked\', \'Connection\': \'keep-alive\', \'Vary\': \'Accept-Encoding\', \'Request-Context\': \'appId=cid-v1:2d2e8e63-272e-4b3c-8598-4ee570a0e70d\', \'api-supported-versions\': \'1.0, 2018-03-01-preview, 2018-11-19\', \'x-ms-client-request-id\': \'32956e8e1d724ae68d37353bcb6f7b3c\', \'x-ms-client-session-id\': \'\', \'Strict-Transport-Security\': \'max-age=15724800; includeSubDomains; preload\', \'Content-Encoding\': \'gzip\'}\nContent: b\'{"code":"NotFound","statusCode":404,"message":"The specified resource was not found","details":[{"code":"NotFoundComputeResource","message":"Compute Resource $(AKS_COMPUTE_NAME) doesn\\\'t exist."}]}\'\n\tInnerException None\n\tErrorResponse \n{\n    "error": {\n        "message": "Received bad response from Model Management Service:\\nResponse Code: 404\\nHeaders: {\'Date\': \'Mon, 03 Feb 2020 20:23:37 GMT\', \'Content-Type\': \'application/json\', \'Transfer-Encoding\': \'chunked\', \'Connection\': \'keep-alive\', \'Vary\': \'Accept-Encoding\', \'Request-Context\': \'appId=cid-v1:2d2e8e63-272e-4b3c-8598-4ee570a0e70d\', \'api-supported-versions\': \'1.0, 2018-03-01-preview, 2018-11-19\', \'x-ms-client-request-id\': \'32956e8e1d724ae68d37353bcb6f7b3c\', \'x-ms-client-session-id\': \'\', \'Strict-Transport-Security\': \'max-age=15724800; includeSubDomains; preload\', \'Content-Encoding\': \'gzip\'}\\nContent: b\'{\\"code\\":\\"NotFound\\",\\"statusCode\\":404,\\"message\\":\\"The specified resource was not found\\",\\"details\\":[{\\"code\\":\\"NotFoundComputeResource\\",\\"message\\":\\"Compute Resource $(AKS_COMPUTE_NAME) doesn\\\\\'t exist.\\"}]}\'"\n    }\n}}\n' }
Reverting to original directory.....

**Comment by algattik** [comment]
The error message shows *Compute Resource $(AKS_COMPUTE_NAME) doesn't exist."* which indicates that the variable AKS_COMPUTE_NAME couldn't be interpolated (that is, it was not defined).

Make sure you follow the steps in docs/getting_started.md, in particular:


Variable Name | Suggested Value
-- | --
AKS_COMPUTE_NAME | aks

**Comment by santhosh340** [comment]
Hi, I have this value.

**Comment by dtzar** [comment]
If AKS_COMPUTE_NAME were properly loading, then you'd see the name listed.  Where did you add it?

**Comment by santhosh340** [comment]
It is in my variable list Library. In addition to addition to this do we need to create AKS cluster before running pipeline or it will provision automatically ?

**Comment by santhosh340** [comment]
It is in my variable list Library. In addition to this do we need to create AKS cluster before running pipeline or it will provision automatically ?

**Comment by dtzar** [comment]
You need to have one provisioned first.  [Sort of mentioned here](https://github.com/microsoft/MLOpsPython/blob/master/docs/getting_started.md#deploy-the-model-to-azure-kubernetes-service).

**Comment by santhosh340** [comment]
I have created one kubernetes cluster with name aks.. and ran the pipeline .. got the similar issue.. I don't know why

**Comment by dtzar** [comment]
Did you attach AKS to Azure ML?

**Comment by santhosh340** [comment]
Yes i attached aks to azure ml workspace Using the CLI.
az aks show -n myexistingcluster -g myresourcegroup --query id
az ml computetarget attach aks -n myaks -i aksresourceid -g myresourcegroup -w myworkspace

**Comment by santhosh340** [comment]
ERROR: {'Azure-cli-ml Version': '1.0.85', 'Error': WebserviceException:
	Message: Received bad response from Model Management Service:
Response Code: 400
Headers: {'Date': 'Tue, 04 Feb 2020 00:16:27 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Request-Context': 'appId=cid-v1:2d2e8e63-272e-4b3c-8598-4ee570a0e70d', 'api-supported-versions': '1.0, 2018-03-01-preview, 2018-11-19', 'x-ms-client-request-id': '2bdb78d8c0f74e11ac0aaca8eaaf891d', 'x-ms-client-session-id': '', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains; preload'}
Content: b'{"code":"BadRequest","statusCode":400,"message":"The request is invalid","details":[{"code":"ComputeResourceNotCreated","message":"Compute resource with Id: aks is not in Succeeded state. Compute provisioning state: Failed"}]}'
	InnerException None
	ErrorResponse 
{
    "error": {
        "message": "Received bad response from Model Management Service:\nResponse Code: 400\nHeaders: {'Date': 'Tue, 04 Feb 2020 00:16:27 GMT', 'Content-Type': 'application/json', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Request-Context': 'appId=cid-v1:2d2e8e63-272e-4b3c-8598-4ee570a0e70d', 'api-supported-versions': '1.0, 2018-03-01-preview, 2018-11-19', 'x-ms-client-request-id': '2bdb78d8c0f74e11ac0aaca8eaaf891d', 'x-ms-client-session-id': '', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains; preload'}\nContent: b'{\"code\":\"BadRequest\",\"statusCode\":400,\"message\":\"The request is invalid\",\"details\":[{\"code\":\"ComputeResourceNotCreated\",\"message\":\"Compute resource with Id: aks is not in Succeeded state. Compute provisioning state: Failed\"}]}'"
    }
}}
Output:  
##[error]Error Code: [1]
##[error]Error: Failed to deploy model(s)
{ code: 1,
  error: undefined,
  stdout: '',
  stderr: 'ERROR: {\'Azure-cli-ml Version\': \'1.0.85\', \'Error\': WebserviceException:\n\tMessage: Received bad response from Model Management Service:\nResponse Code: 400\nHeaders: {\'Date\': \'Tue, 04 Feb 2020 00:16:27 GMT\', \'Content-Type\': \'application/json\', \'Transfer-Encoding\': \'chunked\', \'Connection\': \'keep-alive\', \'Request-Context\': \'appId=cid-v1:2d2e8e63-272e-4b3c-8598-4ee570a0e70d\', \'api-supported-versions\': \'1.0, 2018-03-01-preview, 2018-11-19\', \'x-ms-client-request-id\': \'2bdb78d8c0f74e11ac0aaca8eaaf891d\', \'x-ms-client-session-id\': \'\', \'Strict-Transport-Security\': \'max-age=15724800; includeSubDomains; preload\'}\nContent: b\'{"code":"BadRequest","statusCode":400,"message":"The request is invalid","details":[{"code":"ComputeResourceNotCreated","message":"Compute resource with Id: aks is not in Succeeded state. Compute provisioning state: Failed"}]}\'\n\tInnerException None\n\tErrorResponse \n{\n    "error": {\n        "message": "Received bad response from Model Management Service:\\nResponse Code: 400\\nHeaders: {\'Date\': \'Tue, 04 Feb 2020 00:16:27 GMT\', \'Content-Type\': \'application/json\', \'Transfer-Encoding\': \'chunked\', \'Connection\': \'keep-alive\', \'Request-Context\': \'appId=cid-v1:2d2e8e63-272e-4b3c-8598-4ee570a0e70d\', \'api-supported-versions\': \'1.0, 2018-03-01-preview, 2018-11-19\', \'x-ms-client-request-id\': \'2bdb78d8c0f74e11ac0aaca8eaaf891d\', \'x-ms-client-session-id\': \'\', \'Strict-Transport-Security\': \'max-age=15724800; includeSubDomains; preload\'}\\nContent: b\'{\\"code\\":\\"BadRequest\\",\\"statusCode\\":400,\\"message\\":\\"The request is invalid\\",\\"details\\":[{\\"code\\":\\"ComputeResourceNotCreated\\",\\"message\\":\\"Compute resource with Id: aks is not in Succeeded state. Compute provisioning state: Failed\\"}]}\'"\n    }\n}}\n' }
Reverting to original directory.....
/__w/1/s

**Comment by santhosh340** [comment]
I am following the same as given here- https://github.com/microsoft/MLOpsPython/blob/master/docs/getting_started.md
Is this updated one or if not please provide the step by step process.

**Comment by santhosh340** [comment]
I am following the same as given here- https://github.com/microsoft/MLOpsPython/blob/master/docs/getting_started.md
Is this updated one or please provide the step by step process.

**Comment by algattik** [comment]
> I have created one kubernetes cluster with name aks.. and ran the pipeline .. got the similar issue.. I don't know why

The name you need to provide is the one used to attach the cluster, not the name of the cluster itself. I.e. it is the `-n` parameter in `az ml computetarget attach aks -n myaks -i aksresourceid -g myresourcegroup -w myworkspace`

But in any case the problem is not with the value of the variable - since it's not being interpolated at all, the issue is probably one of those:
- The variable group name doesn't match the name in the pipeline
- The name of the variable in the variable group doesn't match

Maybe try adding the variable to `.pipelines/diabetes_regression-variables.yml` as a temporary workaround and test.

**Comment by santhosh340** [comment]
Is there any update on this?

**Comment by dtzar** [comment]
@santhosh340 - @algattik is right.  Did you try his steps?  Deployments are still working fine to Web, ACI, and AKS on our pipeline so it is likely an issue on your side.

**Comment by santhosh340** [comment]
I verified above steps, both resource group and variables are correct. But still the same issue, i don't know whats the issue here but i will try to do from scratch tmr.
Thankss

**Comment by santhosh340** [comment]
> > I have created one kubernetes cluster with name aks.. and ran the pipeline .. got the similar issue.. I don't know why
> 
> The name you need to provide is the one used to attach the cluster, not the name of the cluster itself. I.e. it is the `-n` parameter in `az ml computetarget attach aks -n myaks -i aksresourceid -g myresourcegroup -w myworkspace`
> 
> But in any case the problem is not with the value of the variable - since it's not being interpolated at all, the issue is probably one of those:
> 
> * The variable group name doesn't match the name in the pipeline
> * The name of the variable in the variable group doesn't match
> 
> Maybe try adding the variable to `.pipelines/diabetes_regression-variables.yml` as a temporary workaround and test.

Is that your problem solved?

**Comment by gdprtcl** [comment]
Hi @santhosh340 I would request you to go back to you Azure ML Workspace.
>> Assets >> Compute >> Inference Clusters >> +New
This will allow you to create a new ML Compute target (AKS Cluster).
1. Use this name as the value to AKS_COMPUTE_NAME.
2. Use the name of the AKS service just created in your Resource Group as the value to AKS_DEPLOYMENT_NAME.

**Comment by santhosh340** [comment]
> Hi @santhosh340 I would request you to go back to you Azure ML Workspace.
> 
> > > Assets >> Compute >> Inference Clusters >> +New
> > > This will allow you to create a new ML Compute target (AKS Cluster).
> 
> 1. Use this name as the value to AKS_COMPUTE_NAME.
> 2. Use the name of the AKS service just created in your Resource Group as the value to AKS_DEPLOYMENT_NAME.

I have given AKS_COMPUTE_NAME as aks cluster name and AKS_DEPLOYMENT_NAME as some x name. Azure ML Model Deploy stage succeeded and got new issue  Smoke test failed.
Traceback (most recent call last):
  File "/usr/local/envs/mlopspython_ci/lib/python3.7/site-packages/urllib3/connection.py", line 157, in _new_conn
    (self._dns_host, self.port), self.timeout, **extra_kw
  File "/usr/local/envs/mlopspython_ci/lib/python3.7/site-packages/urllib3/util/connection.py", line 84, in create_connection
    raise err
  File "/usr/local/envs/mlopspython_ci/lib/python3.7/site-packages/urllib3/util/connection.py", line 74, in create_connection
    sock.connect(sa)
TimeoutError: [Errno 110] Connection timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/envs/mlopspython_ci/lib/python3.7/site-packages/urllib3/connectionpool.py", line 672, in urlopen
    chunked=chunked,
  File "/usr/local/envs/mlopspython_ci/lib/python3.7/site-packages/urllib3/connectionpool.py", line 387, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/envs/mlopspython_ci/lib/python3.7/http/client.py", line 1252, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/local/envs/mlopspython_ci/lib/python3.7/http/client.py", line 1298, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/local/envs/mlopspython_ci/lib/python3.7/http/client.py", line 1247, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/local/envs/mlopspython_ci/lib/python3.7/http/client.py", line 1026, in _send_output
    self.send(msg)
  File "/usr/local/envs/mlopspython_ci/lib/python3.7/http/client.py", line 966, in send
    self.connect()
  File "/usr/local/envs/mlopspython_ci/lib/python3.7/site-packages/urllib3/connection.py", line 184, in connect
    conn = self._new_conn()
  File "/usr/local/envs/mlopspython_ci/lib/python3.7/site-packages/urllib3/connection.py", line 169, in _new_conn
    self, "Failed to establish a new connection: %s" % e
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f59c7bf8b50>: Failed to establish a new connection: [Errno 110] Connection timed out



Any thoughts on above error.. 

Thanks

**Comment by gdprtcl** [comment]
> I have given AKS_COMPUTE_NAME as aks cluster name and AKS_DEPLOYMENT_NAME as some x name.

This is not how you should configure. The doc states, "Set AKS_COMPUTE_NAME to the Compute name of the Inference Cluster that references the Azure Kubernetes Service cluster in your Azure ML Workspace."
AKS_DEPLOYMENT_NAME will be the AKS service which gets deployed in your Resource Group. 

Try in the correct manner and then let me know what you get.

**Comment by santhosh340** [comment]
> > I have given AKS_COMPUTE_NAME as aks cluster name and AKS_DEPLOYMENT_NAME as some x name.
> 
> This is not how you should configure. The doc states, "Set AKS_COMPUTE_NAME to the Compute name of the Inference Cluster that references the Azure Kubernetes Service cluster in your Azure ML Workspace."
> AKS_DEPLOYMENT_NAME will be the AKS service which gets deployed in your Resource Group.
> 
> Try in the correct manner and then let me know what you get.
> 
> @santhosh340

Hi @gdprtcl I changed as above... still i am getting the same above smoke test issue...

**Comment by santhosh340** [comment]
@gdprtcl 
![Screen Shot 2020-03-11 at 10 59 06 AM](https://user-images.githubusercontent.com/39865200/76437416-86b58800-6387-11ea-9fe3-7a2c0538f5be.png)

**Comment by santhosh340** [comment]
@gdprtcl @dtzar @algattik 
Hi, based on suggestion from users and team I am using private AKS cluster. Before starting MLOPS training pipeline i need to attach aks cluster to ML workspace. My attachment status is getting failed, it says 
{
  "location": "eastus2",
  "name": "aks-mlops-poc",
  "provisioningErrors": [
    {
      "error": {
        "code": "InternalServerError",
        "message": "An internal server error occurred. If the problem persists, contact support."
      }
    }
  ],
  "provisioningState": "Failed"
}.

Can we able to attach private aks cluster to machine learning workspace??

**Comment by sudivate** [comment]
@santhosh340 
Unfortunately, Azure ML service does not support attaching the private AKS cluster to AML workspace.

**Comment by santhosh340** [comment]
> @santhosh340
> Unfortunately, Azure ML service does not support attaching the private AKS cluster to AML workspace.

@sudivate Its not completed

**Comment by algattik** [comment]
It should work with
```
az ml computetarget create aks --load-balancer-type InternalLoadBalancer --load-balancer-subnet $SUBNET ...
```

**Comment by klipare** [comment]
> @Santhosh340
> Unfortunately, Azure ML service does not support attaching the private AKS cluster to AML workspace.

Can you support this claim with a document? Please provide the evidence also.

**Comment by JDouss** [comment]
Hi, is there any update on this topic? I am also having this problem
